types/src/transaction/webauthn.rs

```diff
@@ -0,0 +1,707 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::transaction::authenticator::AnyPublicKey;
+use anyhow::{anyhow, Result};
+use aptos_crypto::{
+    hash::CryptoHash, secp256r1_ecdsa, signing_message, CryptoMaterialError, HashValue, Signature,
+};
+use passkey_types::{crypto::sha256, webauthn::CollectedClientData, Bytes};
+use serde::{Deserialize, Serialize};
+
+/// Returns the binary concatenation of
+/// 1. [`authenticator_data_bytes`](PartialAuthenticatorAssertionResponse) and
+/// 2. SHA-256 hash of [`client_data_json`](PartialAuthenticatorAssertionResponse),
+///
+/// See <https://www.w3.org/TR/webauthn-3/#sctn-verifying-assertion>
+fn generate_verification_data(authenticator_data_bytes: &[u8], client_data_json: &[u8]) -> Vec<u8> {
+    // Let hash be the result of computing a hash over the clientData using SHA-256.
+    let client_data_json_hash = sha256(client_data_json);
+    // Binary concatenation of authData and hash.
+    // Note: This is compatible with signatures generated by FIDO U2F
+    // authenticators. See §6.1.2 FIDO U2F Signature Format Compatibility
+    // See <https://www.w3.org/TR/webauthn-3/#sctn-fido-u2f-sig-format-compat>
+    [authenticator_data_bytes, &client_data_json_hash]
+        .concat()
+        .to_vec()
+}
+
+/// This checks that the SHA3-256 of the [`signing_message(message)`](signing_message)
+/// is equal to the actual `challenge` from `client_data_json` in `PartialAuthenticatorAssertionResponse`
+fn verify_expected_challenge_from_message_matches_actual<T: Serialize + CryptoHash>(
+    message: &T,
+    actual_challenge: &[u8],
+) -> std::result::Result<(), CryptoMaterialError> {
+    // Generate signing_message, which is the BCS encoded bytes of message, prefixed with a hash
+    let signing_message_bytes = signing_message(message)?;
+    // Expected challenge is SHA3-256 digest of RawTransaction bytes
+    let expected_challenge = HashValue::sha3_256_of(signing_message_bytes.as_slice());
+
+    expected_challenge
+        .to_vec()
+        .as_slice()
+        .eq(actual_challenge)
+        .then_some(())
+        .ok_or(CryptoMaterialError::ValidationError)
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, Serialize, Deserialize)]
+pub enum WebAuthnSignature {
```

Comment:
I propose naming this AuthenticatorSignature.  the 'WebAuthn' namespace is already crowded and its quite confusing for an AnySignature of Signature::WebAuthn contain a PAAR that contains a WebAuthnSignature

AuthenticatorSignature makes a lot of sence to me because it is a signature that results from the authenticator's signing

Reply:
I think the challenge is that we also use Authenticator everywhere as well (AccountAuthenticator, TransactionAuthenticator, etc). Do you have another name for it perhaps? Otherwise I think I might keep it as is

Reply:
AssertionSignature?

Reply:
Ok that works